### PR TITLE
fix the error prone account creation method of blob disk

### LINKED ISSUE #59738: fix the error prone account creation method of blob disk
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

> Uncomment only one, leave it on its own line: 
>
> /kind bug
> /kind feature

**What happened**:
Original account creation method of blob disk would create account name like `pvc388263479001`, `pvc388263479002` which is error prone, need to use a new way to generate new storage accounts for blob disk

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig azure
/assign

### PR DESCRIPTION
**Pull request by andyzhangx**
**What this PR does / why we need it**:
use new account generation method for blob disk to fix the error prone account creation method of blob disk

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #59738

**Special notes for your reviewer**:

**Release note**:

```
fix the error prone account creation method of azure blob disk
```

/assign @karataliu 
/sig azure

### CODE REVIEW

**Review by feiskyer — APPROVED** [review]
LGTM

File: pkg/cloudprovider/providers/azure/azure_blobDiskController.go [hunk 1]
```diff
@@ -555,7 +543,7 @@ func (c *BlobDiskController) findSANameForDisk(storageAccountType storage.SkuNam
 	countAccounts := 0 // account of this type.
 	for _, v := range c.accounts {
 		// filter out any stand-alone disks/accounts
-		if strings.Index(v.name, storageAccountNameMatch) != 0 {
+		if strings.Index(v.name, azureDiskAccountNamePrefix) != 0 {
```
Comment (karataliu) [inline pkg/cloudprovider/providers/azure/azure_blobDiskController.go]: !strings.HasPrefix better

File: pkg/cloudprovider/providers/azure/azure_blobDiskController.go [hunk 2]
```diff
@@ -301,17 +297,9 @@ func (c *BlobDiskController) DeleteBlobDisk(diskURI string) error {
...
-	hash := MakeCRC32(uniqueString)
-	//used to generate a unique container name used by this cluster PVC
-	defaultContainerName = hash
-
-	storageAccountNamePrefix = fmt.Sprintf(storageAccountNameTemplate, hash)
-	// Used to filter relevant accounts (accounts used by shared PVC)
-	storageAccountNameMatch = storageAccountNamePrefix
-	// Used as a template to create new names for relevant accounts
-	storageAccountNamePrefix = storageAccountNamePrefix + "%s"
+	defaultContainerName = MakeCRC32(c.common.resourceGroup + c.common.location + c.common.subscriptionID)
```
Comment (karataliu) [inline pkg/cloudprovider/providers/azure/azure_blobDiskController.go]: Any reason we need a unique container name? What about using a fixed one

File: pkg/cloudprovider/providers/azure/azure_blobDiskController.go [hunk 3]
```diff
@@ -481,7 +469,7 @@ func (c *BlobDiskController) getAllStorageAccounts() (map[string]*storageAccount
 
 	accounts := make(map[string]*storageAccountState)
 	for _, v := range *accountListResult.Value {
-		if strings.Index(*v.Name, storageAccountNameMatch) != 0 {
+		if strings.Index(*v.Name, azureDiskAccountNamePrefix) != 0 {
```
Comment (karataliu) [inline pkg/cloudprovider/providers/azure/azure_blobDiskController.go]: The code accesses *v.Name first, and check v.Name == nil in next block? Shall fix that?

File: pkg/cloudprovider/providers/azure/azure_storage.go [hunk 4]
```diff
@@ -27,6 +27,7 @@ import (
 const (
 	defaultStorageAccountType  = string(storage.StandardLRS)
 	fileShareAccountNamePrefix = "f"
+	azureDiskAccountNamePrefix = "dk"
```
Comment (karataliu) [inline pkg/cloudprovider/providers/azure/azure_storage.go]: why not 'd', if 'k' has special meaning consider add some comments here

File: pkg/cloudprovider/providers/azure/azure_blobDiskController.go (hunk 3 above)
Comment (andyzhangx) [reply pkg/cloudprovider/providers/azure/azure_blobDiskController.go]: Good catch. fixed.

File: pkg/cloudprovider/providers/azure/azure_blobDiskController.go (hunk 1 above)
Comment (andyzhangx) [reply pkg/cloudprovider/providers/azure/azure_blobDiskController.go]: fixed

File: pkg/cloudprovider/providers/azure/azure_storage.go (hunk 4 above)
Comment (andyzhangx) [reply pkg/cloudprovider/providers/azure/azure_storage.go]: This is for support kind = Shared accounts. one letter `d` is not enough, and I also don't want to use `disk` it has 4 letters, so I use `dk` it's not a word, to differentiate customer self created accounts. A little balancing tricky ...

File: pkg/cloudprovider/providers/azure/azure_blobDiskController.go (hunk 2 above)
Comment (andyzhangx) [reply pkg/cloudprovider/providers/azure/azure_blobDiskController.go]: just keep it, it's not related to this PR, I just used the original code.
unique or fixed name, all looks good.

**Review by andyzhangx — COMMENTED** [review]
@karataliu PTAL again.

**Comment by andyzhangx** [comment]
@karataliu now below are the prefix nameing:
```
+	fileShareAccountNamePrefix     = "f"
 +	sharedDiskAccountNamePrefix    = "ds"
 +	dedicatedDiskAccountNamePrefix = "dd"
```

**Comment by andyzhangx** [comment]
@karataliu I added a new common func `SearchStorageAccount` for CreateFileShare and CreateVolume funcs, this PR should be the last of series of PRs which is targeting to refactor k8s volume on azure code stating from 2017.9, I will cherry pick this PR to v1.8, v1.9. Thanks for your code review.

File: pkg/cloudprovider/providers/azure/azure_storageaccount.go [hunk 5]
```diff
@@ -75,3 +79,52 @@ func (az *Cloud) getStorageAccesskey(account string) (string, error) {
 	}
 	return "", fmt.Errorf("no valid keys")
 }
+
+// SearchStorageAccount search storage account, create one storage account(with genAccountNamePrefix) if not found, return accountName, accountKey
+func (az *Cloud) SearchStorageAccount(accountName, accountType, location, genAccountNamePrefix string) (string, string, error) {
```
Comment (karataliu) [inline pkg/cloudprovider/providers/azure/azure_storageaccount.go]: This func is to be used in package only, consider making it lower case

Also 'search' looks like a read only operation, may be use 'ensure' better, like other places

File: pkg/cloudprovider/providers/azure/azure_storage.go [hunk 6]
```diff
@@ -20,57 +20,19 @@ import (
...
-	"github.com/Azure/go-autorest/autorest/to"
 	"github.com/golang/glog"
 )
 
 const (
-	defaultStorageAccountType  = string(storage.StandardLRS)
-	fileShareAccountNamePrefix = "f"
+	defaultStorageAccountType      = string(storage.StandardLRS)
+	fileShareAccountNamePrefix     = "f"
+	sharedDiskAccountNamePrefix    = "ds"
```
Comment (karataliu) [inline pkg/cloudprovider/providers/azure/azure_storage.go]: Those should be placed in the file where it is in use? 
azure_storage.go now contains file-service only code, maybe we need to merge it to azure_file.go someday.

File: pkg/cloudprovider/providers/azure/azure_storage.go [hunk 7]
```diff
@@ -20,57 +20,19 @@ import (
...
-			err := <-errchan
-			if err != nil {
-				return "", "", fmt.Errorf(fmt.Sprintf("Failed to create storage account %s, error: %s", accountName, err))
-			}
-		}
-	}
-
-	// find the access key with this account
-	accountKey, err := az.getStorageAccesskey(accountName)
+	accountName, accountKey, err := az.SearchStorageAccount(accountName, accountType, location, fileShareAccountNamePrefix)
```
Comment (karataliu) [inline pkg/cloudprovider/providers/azure/azure_storage.go]: We'd better not assign value to 'accountName' here.

Consider the code in SearchStorageAccount:
```
+	accountKey, err := az.getStorageAccesskey(accountName)
 +	if err != nil {
 +		return "", "", fmt.Errorf("could not get storage key for storage account %s: %v", accountName, err)
 +	}
```

If user specifies 'AccountName', but later getStorageAccesskey fails, then SearchStorageAccount returns empty for first parameter.

In this case the follow-up error reported would use empty value for accountName, which does not sound good to user.

Also for CreateVolume.

File: pkg/cloudprovider/providers/azure/azure_blobDiskController.go [hunk 8]
```diff
@@ -555,7 +516,7 @@ func (c *BlobDiskController) findSANameForDisk(storageAccountType storage.SkuNam
 	countAccounts := 0 // account of this type.
 	for _, v := range c.accounts {
 		// filter out any stand-alone disks/accounts
-		if strings.Index(v.name, storageAccountNameMatch) != 0 {
+		if strings.Index(v.name, sharedDiskAccountNamePrefix) != 0 {
```
Comment (karataliu) [inline pkg/cloudprovider/providers/azure/azure_blobDiskController.go]: Also strings.HasPrefix ?

File: pkg/cloudprovider/providers/azure/azure_storage.go (hunk 7 above)
Comment (andyzhangx) [reply pkg/cloudprovider/providers/azure/azure_storage.go]: good catch, fixed both funcs.

File: pkg/cloudprovider/providers/azure/azure_storageaccount.go (hunk 5 above)
Comment (andyzhangx) [reply pkg/cloudprovider/providers/azure/azure_storageaccount.go]: fixed, I am also considering a good name, ensureStorageAccount is better

**Review by andyzhangx — COMMENTED** [review]
@karataliu  really good comments, PTAL.

**Review by karataliu — APPROVED** [review]
/lgtm

**Pull request merged**